app/src/main/java/com/keylesspalace/tusky/components/compose/dialog/FocusDialog.kt

```diff
@@ -0,0 +1,235 @@
+
+    // From Glide: "We now own this Bitmap. It's our responsibility to replace it in the pool outside this method
+    // when we're finished with it."
+    return argbBitmap
+}
+
+// Private, but necessary to implement BitmapTransformation, function extracted from Glide
+private fun getAlphaSafeConfig(inBitmap: Bitmap): Bitmap.Config {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        // Avoid short circuiting the sdk check.
+        if (Bitmap.Config.RGBA_F16 == inBitmap.config) { // NOPMD
+            return Bitmap.Config.RGBA_F16
+        }
+    }
+    return Bitmap.Config.ARGB_8888
+}
+
+private val transparentDarkGray = 0x40000000
+
+/** Glide BitmapTransformation which overlays a highlight on a focus point. */
+class HighlightFocus(val focus: Focus) : BitmapTransformation() {
+    override fun transform(pool: BitmapPool, inBitmap: Bitmap, outWidth: Int, outHeight: Int): Bitmap {
+        // Draw overlaid target
+        val bitmapDrawableLock: Lock = TransformationUtils.getBitmapDrawableLock()
+        val safeConfig: Bitmap.Config = getAlphaSafeConfig(inBitmap)
+        val toTransform: Bitmap = getAlphaSafeBitmap(pool, inBitmap)
+        val result = pool[toTransform.width, toTransform.height, safeConfig]
+
+        val plainPaint = Paint()
+        val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG)
+        strokePaint.setAntiAlias(true)
+        strokePaint.setStyle(Paint.Style.STROKE)
+        val strokeWidth = 10.0f
+        strokePaint.setStrokeWidth(strokeWidth)
+        strokePaint.setColor(Color.WHITE)
+        val curtainPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+        curtainPaint.style = Paint.Style.FILL
+        curtainPaint.color = transparentDarkGray
+
+        bitmapDrawableLock.lock()
+        try {
+            val canvas: Canvas = Canvas(result)
+
+            canvas.drawBitmap(toTransform, Matrix.IDENTITY_MATRIX, plainPaint)
+
+            // Canvas range is 0..size Y-down but Mastodon API range is -1..1 Y-up
+            val width = result.width.toFloat()
+            val height = result.height.toFloat()
+            val x = (focus.x + 1.0f) / 2.0f * width
+            val y = (1.0f - focus.y) / 2.0f * height
+            val circleRadius = Math.min(width, height).toFloat() / 4.0f
+
+            val curtainPath = Path() // Draw a flood fill with a hole cut out of it
+            curtainPath.setFillType(Path.FillType.WINDING)
+            curtainPath.addRect(0.0f, 0.0f, width, height, Path.Direction.CW)
+            curtainPath.addCircle(x, y, circleRadius, Path.Direction.CCW)
+            canvas.drawPath(curtainPath, curtainPaint)
+
+            canvas.drawCircle(x, y, circleRadius, strokePaint)
+            canvas.drawCircle(x, y, strokeWidth / 2.0f, strokePaint)
+
+            canvas.setBitmap(null)
+        } finally {
+            bitmapDrawableLock.unlock()
+        }
+
+        if (!toTransform.equals(inBitmap)) {
+            pool.put(toTransform)
```

Comment:
I am very, VERY confused if this is desirable or not. I *think* it's good because this will cache the image with the circle/shadow drawn on so if the set focus dialog is opened a second time it will open faster.